modules/storages/app/services/storages/health/health_service.rb

```diff
@@ -0,0 +1,76 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2024 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  module Health
+    class HealthService
+      def initialize(storage:)
+        @storage = storage
+      end
+
+      def healthy
+        was_unhealthy = @storage.health_unhealthy?
+        reason = @storage.health_reason
+
+        @storage.mark_as_healthy
+
+        admin_users.each do |admin|
+          ::Storages::StoragesMailer.notify_healthy(admin, @storage, reason).deliver_later if was_unhealthy
```

Comment:
We have a check inside `Storages::Storage#mark_as_healthy` if it was healthy before. For me it seems like a duplication. Does not it?

Reply:
The `HealthService` itself is now in charge of setting the state on the storage. I know that `mark_as_healthy` is doing a check, but it is too transparent from the outside to see, if it was unhealthy before. I need this to determine if I need to send mails or not. 

We could think about something like a `finalize` method to call a block if it has changed or something like that. But for now I think this would just make it more complicated.

My favorite thing would be, to extract this logic from the model completely, because it's not the right this to handle this invariant inside a data class, which main purpose should be writing consistent data to the database an obtaining it again.